rfd/0078-trait-transforms.md

```diff
@@ -0,0 +1,754 @@
+---
+authors: Nic Klaassen (nic@goteleport.com)
+state: draft
+---
+
+# RFD 78 - Trait Transforms
+
+## What
+
+A new feature which allows admins to filter, transform, and extend incoming
+SAML assertions and OIDC claims from their identity provider before they are
+embedded in the traits of their Teleport user's certificates.
+
+## Why
+
+Teleport admins often do not have control over their company's SSO provider and
+which claims will be provided to Teleport.
+All of these claims will be embedded in each user's Teleport certificate as
+`traits`.
+
+SSO providers can include hundreds to thousands of claims which are not
+necessary or used by Teleport at all, unecessarily bloating the certificate size
+of all users.
+This can adversly impact latency and throughput of Teleport operations, and in
+an extreme case could reach a limit which would render Teleport unusable with
+that identity provider.
+Trait Transforms will allow the Teleport admin to filter out claims which are
+unecessary before they make it into the Teleport certificate.
+
+Admins often wish to extend an incoming claim with extra values which make sense
+for their Teleport deployment.
+For example, users within the `splunk` group should also be added to the `dbs`
+group within Teleport so that they can access databases.
+Trait Transforms will allow the Teleport admins to modify existing traits and
+add new ones to solve these problems.
+
+If teleport admins cannot add or change the claims provided by their SSO
+provider, making simple rules like this can be extremely cumbersome or
+impossible within teleport.
+There are currently teleport users with deployments which automatically generate
+hundreds to thousands of roles to get around these limitations.
+
+## Details
+
+Example trait transform spec:
+
+```yaml
+kind: trait_transform
+version: v1
+metadata:
+  name: example
+spec:
+  # priority can be used to order the evaluation of multiple trait transforms in
+  # a cluster. Lower priorities will be evaluated first. Trait transforms with
+  # the same priority will be ordered by a lexicographical sort by their names. 
+  priority: 0
```

Comment:
I would not worry about that so early on. Otherwise we'd have to use compare and swap or other locking mechanism to prevent duplicates on create.

Reply:
yeah, good call

Reply:
Oh I hadn't seen your comment sasha, GitHub was out of date. I had updated the RFD to block dupes, but reverted now.

Compare-and-swap wouldn't be too difficult for a single resource, but across multiple resources it could be a bit of a pain.